[SpeechSynthesis] Emit error event for all cancelled utterances

pgorszkowski-igalia opened the pull request:
#### 5a4c608c84884c202ff8cedbab92bd9fe15b6204
<pre>
[SpeechSynthesis] Emit error event for all cancelled utterances
<a href="https://bugs.webkit.org/show_bug.cgi?id=290750">https://bugs.webkit.org/show_bug.cgi?id=290750</a>

Reviewed by Sihui Liu.

Dispatch an error event &quot;canceled&quot; for queued utterances.

Specification: <a href="https://dvcs.w3.org/hg/speech-api/raw-file/tip/webspeechapi">https://dvcs.w3.org/hg/speech-api/raw-file/tip/webspeechapi</a>, point: 5.2.5.1.

Original author: Andrzej Surdej &lt;Andrzej_Surdej@comcast.com&gt;

See: <a href="https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1474">https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1474</a>

* LayoutTests/fast/speechsynthesis/speech-synthesis-cancel-expected.txt:
* LayoutTests/fast/speechsynthesis/speech-synthesis-cancel.html:
* Source/WebCore/Modules/speech/SpeechSynthesis.cpp:
(WebCore::SpeechSynthesis::cancel):

Canonical link: <a href="https://commits.webkit.org/293536@main">https://commits.webkit.org/293536@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d3bd916b3e98bcbf19624f2b0fb7787ba323ead6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97884 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17509 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7730 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102998 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48414 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99929 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17803 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25962 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74526 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31709 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100887 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13479 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88457 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54883 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13255 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6389 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47856 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83241 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6462 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105376 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24964 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18181 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83529 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25336 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84638 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82965 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21303 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27599 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5283 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18582 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24925 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30094 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24747 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28061 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26321 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/Modules/speech/SpeechSynthesis.cpp:
```diff
@@ -178,16 +178,28 @@ void SpeechSynthesis::cancel()
+            // Current utterance is handled in platform cancel()
+            if (current.get() != utterance.ptr())
+                utterance.get().errorEventOccurred(eventNames().errorEvent, SpeechSynthesisErrorCode::Canceled);
+        }
```
I think we need to fire event for pending utterances in queue in all cases, no matter the synthesizer is in-process or out-of-process; i.e. this part should be moved out of the `else if m_platformSpeechSynthesizer` condition.

pgorszkowski-igalia committed:
[SpeechSynthesis] Emit error event for all cancelled utterances
https://bugs.webkit.org/show_bug.cgi?id=290750

Reviewed by Sihui Liu.

Dispatch an error event "canceled" for queued utterances.

Specification: https://dvcs.w3.org/hg/speech-api/raw-file/tip/webspeechapi, point: 5.2.5.1.

Original author: Andrzej Surdej <Andrzej_Surdej@comcast.com>

See: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1474

* LayoutTests/fast/speechsynthesis/speech-synthesis-cancel-expected.txt:
* LayoutTests/fast/speechsynthesis/speech-synthesis-cancel.html:
* Source/WebCore/Modules/speech/SpeechSynthesis.cpp:
(WebCore::SpeechSynthesis::cancel):

Canonical link: https://commits.webkit.org/293536@main
